internal/services/controller/controller.go

```diff
@@ -279,7 +282,16 @@ func (c *Controller) Run(ctx context.Context) error {
 		c.healthzProvider.Initialized()
 
 		c.log.Infof("sending cluster deltas every %s", c.cfg.Interval)
-		wait.Until(func() {
+
+		wait.NonSlidingUntil(func() {
+			// Check if another goroutine is trying to send deltas.
+			if !c.sendMu.TryLock() {
+				// If it is, don't try to send deltas on this turn to avoid a backlog of sends.
+				return
+			}
+			// Since Mutex.TryLock() acquires a lock on success,
+			// release it immediately to allow the new sending goroutine to do its job.
+			c.sendMu.Unlock()
```

Comment:
What is the purpose of locking it twice (in the `wait` func and `send` func)? Why couldn't you lock it only in `wait` and use `defer c.sendMu.Unlock()` in loc 294?

Reply:
We could do that. But the purpose of doing this "dry-run" lock is to maintain locality of reasoning and have the function manage its own locking lifecycle.

The `send()` function deals with sending deltas. When you look at the `send()` source code in isolation and see `sendMu` here, you immediately know that only one `send()` is supposed to be running. If you ever need to make any changes to `send()`, you are immediately aware that it needs a lock because it manages its locking on its own.

In turn, `wait.NonSlidingUntil()` deals with scheduling sends. It uses `sendMu` just as a convenient way to check if a `send()` is currently running and skip triggering another one to avoid a growing backlog of `send()`s waiting on the lock. If you need to change scheduling, you are also immediately aware of how it uses the mutex.

Otherwise, if we just lock the mutex in `wait.NonSlidingUntil()`, when you make changes to `send()` you also need to be aware that this function assumes it is locked somewhere else.

I think it makes it simpler to reason about code if you have to jump around the file as little as possible, and doing a somewhat awkward `Unlock()` here is a price I’m okay with.

Reply:
I have an 180 opinion. I think it makes it more complicated because the extra locks aren't needed for functionality and then the reader is left to wonder why it's needed. Whenever dealing with concurrency structures, it's best to keep it as simple as possible. Now it looks like a bug.